### PR TITLE
Add mTLS to kubeadm etcd liveness probe.

### LINKED ISSUE #720: etcd fails kubelet's health checks
## What keywords did you search in kubeadm issues before filing this one?

etcd ssl health

## Is this a BUG REPORT or FEATURE REQUEST?

Choose one: BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`): HEAD

**Environment**: custom/docker-in-docker, docker-in-docker-in-docker
- **Kubernetes version** (use `kubectl version`): HEAD
- **Cloud provider or hardware configuration**: local
- **OS** (e.g. from /etc/os-release): Debian
- **Kernel** (e.g. `uname -a`): 4.9.0-5

## What happened?

tl;dr - kubelet doesn't know the key/crt for etcd, so that etcd returns "non-sense" for its health check.

1. kubeadm creates etcd file
1. kubelet creates etcd container
1. cluster comes up, etcd is taking writes, and everything's great
1. etcd fails kubelet's health checks
1. nothing's great

$ curl -k https://127.0.0.1:2379/health --key apiserver-etcd-client.key --cert apiserver-etcd-client.crt 
{"health": "true"}
$ curl -k https://127.0.0.1:2379/health
curl: (35) error:14094412:SSL routines:ssl3_read_bytes:sslv3 alert bad certificate
$ journalctl | grep "Liveness probe" | grep -v succeeded | tail -n1
Mar 03 09:49:08 7e928d1f665e kubelet[151]: I0303 09:49:08.142334     151 server.go:422] Event(v1.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"etcd-7e928d1f665e", UID:"17c2da92666b4a9242f31873234f3101", APIVersion:"v1", ResourceVersion:"", FieldPath:"spec.containers{etcd}"}): type: 'Warning' reason: 'Unhealthy' Liveness probe failed: Get http://127.0.0.1:2379/health: net/http: HTTP/1.x transport connection broken: malformed HTTP response "\x15\x03\x01\x00\x02\x02"

## What you expected to happen?

I expect etcd to pass health checks.

## How to reproduce it (as minimally and precisely as possible)?

I suspect any kubeadm cluster built from HEAD will have this quirk.

### LINKED ISSUE #59766: [job failure] periodic-kubernetes-e2e-kubeadm-gce-selfhosting
/priority failing-test
/kind bug
/status approved-for-milestone
/sig cluster-lifecycle

https://k8s-testgrid.appspot.com/sig-release-master-blocking#kubeadm-gce-selfhosting
the job is on the [master-blocking dashboard](https://k8s-testgrid.appspot.com/sig-release-master-blocking) and been failing to bring up the cluster.

cc @jdumars @luxas @jessicaochen

### PR DESCRIPTION
**Pull request by stealthybox**
**What this PR does / why we need it**:
We switched etcd over to using mTLS, but the liveness probe is still using http.
Disabling the liveness probe allows etcd to continue operating.

The real fix isn't simple, because we need to generate a client certificate for healthchecking and update the probe to exec `etcdctl` like so: 
https://sourcegraph.com/github.com/coreos/etcd-operator/-/blob/pkg/util/k8sutil/pod_util.go#L71-89

~Working on patching this now.~
This PR now generates the healthcheck identity and updates the liveness probe to use it.

**Which issue(s) this PR fixes**
Fixes #59766
Fixes kubernetes/kubeadm#720

**Special notes for your reviewer**:
We should generate a client cert specifically for etcd health checks so that the apiserver certs can be revoked independently.
This will be stored in `/etc/kubernetes/pki/etcd/` so that we don't have to change the pod's hostMount.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by Q-Lee** [comment]
@krousey this definitely needs to be in the release. The kubeadm tool is simply broken if it's not.

**Comment by stealthybox** [comment]
@kargakis made a 4 line addition to generate the healthcheck cert on etcd upgrade -- please review when you can 👍

**Comment by kargakis** [comment]
flaked on https://github.com/kubernetes/kubernetes/issues/60777
/test pull-kubernetes-e2e-gce

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60728)

File: cmd/kubeadm/app/util/staticpod/utils.go [hunk 1]
```diff
@@ -97,6 +97,24 @@ func ComponentProbe(cfg *kubeadmapi.MasterConfiguration, componentName string, p
 	}
 }
 
+// EtcdProbe is a helper function for building a shell-based, etcdctl v1.Probe object to healthcheck etcd
+func EtcdProbe(cfg *kubeadmapi.MasterConfiguration, componentName string, port int, certsDir string, CACertName string, CertName string, KeyName string) *v1.Probe {
+	tlsFlags := fmt.Sprintf("--cacert=%[1]s/%[2]s --cert=%[1]s/%[3]s --key=%[1]s/%[4]s", certsDir, CACertName, CertName, KeyName)
+	// etcd pod is alive if a linearizable get succeeds.
+	cmd := fmt.Sprintf("ETCDCTL_API=3 etcdctl --endpoints=%s:%d %s get foo", GetProbeAddress(cfg, componentName), port, tlsFlags)
```
Comment (timothysc) [inline cmd/kubeadm/app/util/staticpod/utils.go]: This pre-supposes that etcdctl is installed and it's not part of the installation process.  You'd have to use the client library.

**Review by timothysc — CHANGES_REQUESTED** [review]
Please address.

File: cmd/kubeadm/app/util/staticpod/utils.go (hunk 1 above)
Comment (stealthybox) [reply cmd/kubeadm/app/util/staticpod/utils.go]: Hi Tim,
Would you please clarify? The liveness probe exec's into the container:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#execaction-v1-core

This does assume that `etcdctl` is an included binary on the `PATH` inside the `etcd` container, and that it's capable of using the V3 API.

Comment (stealthybox) [reply cmd/kubeadm/app/util/staticpod/utils.go]: I have verified this works with the current upstream image: `k8s.gcr.io/etcd-amd64:3.1.11`

**Review by timothysc — APPROVED** [review]
/approve

**Pull request merged**